tools/common-subexpr-elim.rkt

```diff
@@ -0,0 +1,67 @@
+#lang racket
+
+(require "common.rkt" "fpcore.rkt")
+
+(provide core-common-subexpr-elim)
+
+(define (core-common-subexpr-elim prog)
+  (match-define (list 'FPCore (list args ...) props ... body) prog)
+  (define cs-hash (count-common-subexpr body))
+  (define cse-body (common-subexpr-elim cs-hash body))
+  `(FPCore ,args ,@props ,cse-body))
+
+(define (combine-common-subexpr-hash h1 h2)
+  (for ([(k v) (in-hash h2)])
+    (if (hash-has-key? h1 k)
+      (hash-set! h1 k #t)
+      (hash-set! h1 k v)))
+  h1)
+
+(define (count-common-subexpr expr)
+  (cond 
+    [(list? expr)
+     (match-define (list op args ...) expr)
+     (let ([cs-hash (make-hash (list (cons expr #f)))])
+       (for ([e args])
+         (combine-common-subexpr-hash cs-hash (count-common-subexpr e)))
+       cs-hash)]
+    [else (make-hash)]))
+
+(define (common-subexpr-elim cs-hash expr)
+  (define intermediates '())
+  (define name-hash (make-hash))
+
+  (define (common-subexpr-body expr)
+    (cond
+      [(list? expr)
+       (match-define (list op args ...) expr)
+       (let ([elimed-exprs (for/list ([arg args])
+                             (common-subexpr-body arg))])
+         (if (hash-ref cs-hash expr #f)
+           (if (hash-has-key? name-hash expr)
+             (hash-ref name-hash expr)
+             (let* ([expr-name (string->symbol
+                                 (format "i~a" (hash-count name-hash)))]
```

Comment:
This will misbehave if one of the arguments to the function is named `i0`. Take a look at `gensym` in `core2c.rkt` line 64 for the way I tend to solve this problem